src/github.com/matrix-org/dendrite/syncserver/storage/current_room_state_table.go

```diff
@@ -39,9 +39,17 @@ const upsertRoomStateSQL = "" +
 const deleteRoomStateByEventIDSQL = "" +
 	"DELETE FROM current_room_state WHERE event_id = $1"
 
+const selectRoomIDsWithMembershipSQL = "" +
+	"SELECT room_id FROM current_room_state WHERE type = 'm.room.member' AND state_key = $1 AND membership = $2"
```

Comment:
You'll probably want an index on current_room_state for this.
Partial indexes might be a neat way to do this https://www.postgresql.org/docs/8.0/static/indexes-partial.html